### PR TITLE
GEOMESA-2812 Add Kryo input that doesn't mutate the underlying buffer

### PR DESCRIPTION
**Pull request by elahrvivaz**
* Space-filling curve impls use Long instead of generics

Signed-off-by: Emilio Lahr-Vivaz <elahrvivaz@ccri.com>

### CODE REVIEW

**Comment by elahrvivaz** [comment]
I'm hopeful that this will fix all those intermittent s2 unit test failures - it seemed to in my local testing.

File: geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/impl/LazyDeserialization.scala [hunk 1]
```diff
@@ -69,7 +70,8 @@ object LazyDeserialization {
         // we create a new kryo input each time, so that position and offset are not affected by other reads
         // this should be thread-safe, as long as the same attribute is not being read in multiple threads
         // (since kryo can mutate the bytes during read)
-        val input = new Input(bytes, offset + (2 * i), length - (2 * i))
+        val input = new NonMutatingInput()
```
Comment (jnh5y) [inline geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/impl/LazyDeserialization.scala]: The comment above should be updated!

File: geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/impl/LazyDeserialization.scala [hunk 2]
```diff
@@ -100,7 +102,9 @@ object LazyDeserialization {
         // we create a new kryo input each time, so that position and offset are not affected by other reads
         // this should be thread-safe, as long as the same attribute is not being read in multiple threads
         // (since kryo can mutate the bytes during read)
-        val input = new Input(bytes, offset + (4 * i), length - (4 * i))
+
```
Comment (jnh5y) [inline geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/impl/LazyDeserialization.scala]: Ditto

File: geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/impl/LazyDeserialization.scala (hunk 1 above)
Comment (elahrvivaz) [reply geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/impl/LazyDeserialization.scala]: thanks, fixed

File: geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/impl/LazyDeserialization.scala (hunk 2 above)
Comment (elahrvivaz) [reply geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/impl/LazyDeserialization.scala]: thanks, fixed

**Comment by jnh5y** [comment]
merged as 346c9e2289d2339f36289adf8e2ccc4bcb577470

**Pull request closed without merging**